blog: frontend asset passthrough (#924)
lyft/clutch: docs/blog/2021-02-01-frontend-asset-passthrough.md

This uniqueness of this content hash allows Clutch to cache bust what the browser has locally allowing the new version to be requested.

<img alt="Problem Diagram" src="https://user-images.githubusercontent.com/2250844/106201546-765bf680-616d-11eb-83d3-c70cf93ba252.png" />

The client makes a request to Clutch and a canary host responds with an `index.html` page with a script tag asking for `main.a3762de8.chunk.js`.
Illustrated by the green and red arrows, there is only one host which has the correct asset the client is asking for, the canary.
If the client's request for `main.a3762de8.chunk.js` goes anywhere else the page will fail to load.

This is not a problem if you are using a CDN, as traditionally there is a backing object storage such as S3 where all of the applications assets are stored.
If a CDN does not have an asset it checks its backing object storage, warms the CDN cache, and serves it to the client.
Alternatively, sticky sessions or session affinity could be used to remediate this problem,
as the requesting client would always be routed to a specific version of the application.

## Solution

This leads us to our solution, which we call Frontend Asset Passthrough.

> How this works is simple: when a request for a static asset does not reside on disk,

Reviewer: ```suggestion
How this works is simple: when a request for a static asset is not found in the binary's embedded filesystem,
```

Let's first take a look at the [proto configuration](https://github.com/lyft/clutch/blob/890245e7d2a1bf91623a9e74b39f1083dbd5ea2c/api/config/gateway/v1/gateway.proto#L105-L119) and then will go into more detail.

```protobuf
message Assets {
  message S3Provider {
    string region = 1;
    string bucket = 2;
    string key = 3;
  }

  oneof provider {
    S3Provider s3 = 1;
  }
}
```

The configuration can be found in the [gatewayOptions](/docs/configuration#gatewayoptions).

At Lyft we utilize AWS S3 as the provider of choice and have implemented this setup as an example;
however, Clutch is extensible and can easily be extended with support for additional blob storage providers.

For S3 the configuration is simple: specify a `region`, `bucket`, and the `key` where the assets live.
This does require you to also configure the `clutch.service.aws` service,
which allows Clutch to fetch these assets via S3 APIs.

```yaml
gateway:
  assets:
    s3:
      region: us-east-1
      bucket: static
      key: clutch

services:
  - name: clutch.service.aws
    typed_config:
      "@type": types.google.com/clutch.config.service.aws.v1.Config
      regions:
        - us-east-1
```

> With this configured, if a user requests an asset that does not reside on disk, Clutch will attempt to serve the file from the configured passthrough provider.

Reviewer: ```suggestion
With this configured, if a user requests an asset that is not embedded in the binary, Clutch will attempt to serve the file from the configured passthrough provider.
```
